sdk/analytics/src/main/java/com/klaviyo/analytics/UserInfo.kt

```diff
@@ -56,9 +70,42 @@ internal object UserInfo {
      */
     var anonymousId: String = ""
         private set(value) { field = persist(ANONYMOUS_ID, value) }
-        get() = field.ifEmpty { fetch(ANONYMOUS_ID, generateUuid).also { anonymousId = it } }
+        get() = field.ifEmpty { fetch(ANONYMOUS_ID, ::generateUuid).also { anonymousId = it } }
+
+    var pushToken: String = ""
+        private set(value) { field = persist(PUSH_TOKEN, value) }
+        get() = field.ifEmpty { fetch(PUSH_TOKEN).also { field = it } }
 
-    private val generateUuid = { UUID.randomUUID().toString() }
+    /**
+     * Track the most recent state of push token + device metadata sent to the backend API
+     */
+    private var pushState: String = ""
+        set(value) { field = persist(PUSH_STATE, value) }
+        get() = field.ifEmpty { fetch(PUSH_STATE).also { field = it } }
+
+    /**
+     * Save push token string to state
+     * If push token or any other device metadata have changed,
+     * invoke the onChanged callback (i.e. to enqueue the API request)
+     */
+    fun setPushToken(token: String, onChanged: () -> Unit) {
+        // Use the request body format as our state tracking value
+        val newPushState = PushTokenApiRequest(token, getAsProfile()).requestBody
+
+        if (newPushState != pushState) {
+            // Optimistic update algorithm: expect request to get to backend,
+            // on failure reset push state (see initializer). The main advantage to
+            // this algorithm is it prevents queueing duplicate requests immediately
+            pushState = newPushState ?: ""
+            pushToken = token
+            onChanged()
+        }
+    }
```

Comment:
Said it earlier, but the comment got lost.
I don't love how state and side effects are being managed, especially with this addition. I had solutions in our very first SDK architecture designs that would separate public API (`Klaviyo` object), state (`UserInfo` object) and side-effects. I would like to fast follow this PR by implementing those designs